### DISCUSSION TITLE
Quarto uses wrong Python version?

### DISCUSSION THREAD

**Discussion by kellertuer** [quarto-dev/quarto-cli#3977] [Q&A] [answered]
I installed quarto last week on two machines and it was working great.
I want to use it with Julia, so I also installed IJulia and Conda and quarto rendered things just fine.

Now it still works on one machine but not on the other and I can not figure out what changed nor what is different.

When I now call

```
❯ quarto render Optimize\!.qmd

Starting julia-1.6 kernel...ERROR: 

[Errno 2] No such file or directory: '/Applications/Julia-1.6.app/Contents/Resources/julia/bin/Julia'
```

which is ok in the sense as _that_ Julia version I deleted in November when I switched to `juliaup` – but _how_ did quarto (and only on one machine) switch to using a different Julia? Or is it maybe using a different Python?
My usual Problem with Python is, that I feel I have far too many Python versions and I never know which version is used where, so any tips in that direction are also welcome.

I have an Apple M1, so my usual Julia I would like to use (and do use on one machine) is the juliaup one – and Python installed via Conda.jl

Any help how to narrow this down appreciated, since I really do not know how quarto switched to trying to use a version that I removed 3 months ago.

**Comment by kellertuer** [comment]
...now I removed said Kernel, then it runs only with Python since there is no Julia left and the (Julia) code in the `.qmd` does not get executed. So It really seems to be the wrong python. But how to tell quarto the right one? I now read through the docs for half an hour but it seems the docs assume, that that all just works?

So _if_ it works its great – if not (due to more than one Python version probably) – I could not find anything yet.

**Comment by cderv** [comment, marked as answer]
Can you run `quarto check jupyter` in the same environment you are using `quarto render` ? 

This should tell you which Python version is seen and used. 

Doc about selection is here : https://quarto.org/docs/computations/python.html#installation. 
From this doc: 

> Quarto will select a version of Python using the [Python Launcher](https://docs.python.org/3/using/windows.html#python-launcher-for-windows) on Windows or system PATH on MacOS and Linux.

So your python found in PATH should be used ? What is on your PATH ? What does `which python3` returns ? 
Using any tooling that put Python on PATH for a project would work with Quarto I believe.

> You can override the version of Python used by Quarto by setting the QUARTO_PYTHON environment variable.

You can always use this environment variable in a project to force Quarto to use a specific version for a project or for your system. 


Hope it helps narrow your configuration and what Quarto sees. Also updating Quarto can help in some situation as we fix issues over time.

**Reply by kellertuer** [reply]
I basically removed quarto and IJulia and installed both again, that helped. My path is crowded, since most things I installed add something. I might have to clean that at some point as well. For now the reinstall and setting the `QUARTO_PATH` resolved the problem.

**Comment by mcaruel** [comment]
Hello,
What is the procedure to change the QUARTO_PYTHON environment variable? I want Quarto to use a specific conda environment not located in the directory of the .qmd document to render.

**Reply by cderv** [reply]
`QUARTO_PYTHON` is an environment variable. You need to set it on your system or in the current session, before `quarto` is called. 
How to set it depend on your OS or tool you are using. 

The procedure is: 

1. Configure `QUARTO_PYTHON` environment variable
2. Run `quarto check` -> it should use the set python

**Reply by delapazruiz** [reply]
Hello @cderv 

Could you provide a file example that defines a Python or R version? 
I am reading the following link, but I'm not familiar with those configurations, and I can not make it work:
https://quarto.org/docs/projects/environment

What file type is the '_environment' mentioned in the above link? Should I write it as a .txt file?
If I want to define an r version, is it correct to type the following in the _environment file?

# QUARTO_R
R_VERSION=4.0.3

I appreciate any help you can provide.

**Reply by mcanouil** [reply]
`_environment` is the name of the file which is a plain text file in which you can set `QUARTO_R=/opt/R/4.2.3/bin`, etc.

**Reply by delapazruiz** [reply]
@mcanouil, Thank you. 

After creating a quarto project with an `renv`, I got the renv.lock file. In this file, I found the following code lines:

  "R": {
    "Version": "4.3.1",
    "Repositories": [
      {
        "Name": "CRAN",
        "URL": "https://packagemanager.posit.co/cran/latest"
      }
    ]
  },
  "Python": {
    "Version": "3.9.15",
    "Type": "virtualenv",
    "Name": "./renv/python/virtualenvs/renv-python-3.9"
  },

I suppose that using `renv` is also a solution to define my preferred Python and R versions, right?

**Reply by mcanouil** [reply]
What does `renv` has to do with setting R/Python for Quarto?
`renv` does not set anything, it only records versions as documented on the tool website (<https://rstudio.github.io/renv/index.html>).

**Reply by delapazruiz** [reply]
The first time, when creating a new quarto project and adding the `renv` option in Rstudio, I got the option to set the R version. It only happens once. Further versions are kept as you choose the first time.

Thank you. I was finally able to define any R version that I wanted. In the _environment file, I typed the following line (for Windows 11): 
` QUARTO_R="C:\Program Files\R\R-4.3.0\bin" `

To confirm that the quarto project was using the version of my choice, I rendered the following .qmd:

---
title: "testRpyversion"
---

```{r}
R.version.string
```

```{python}
import platform
print(platform.python_version())
```

Based on this discussion: <https://github.com/quarto-dev/quarto-cli/issues/6162> I realized that only rendered .qmd files confirm whether your quarto project is using the R/Python version of your choice. Running cell chucks only gives you the versions used by the system, not your quarto project.
